e2e/storage: parameterize container images

pohly opened the pull request:
**What this PR does / why we need it**:

The CSI integration test for hostpath was hard-coded to use the latest
stable release of the sidecar and hostpath container images. This
makes sense for regression testing of changes made in Kubernetes
itself, but the same test is also useful for testing the "canary"
images on quay.io before tagging them as a new release or for testing
locally produced images. Both is now possible via command line
parameters.

**Which issue(s) this PR fixes**:
Related-to: kubernetes-csi/docs#23

**Special notes for your reviewer**:

The commit message has usage instructions.

```release-note
NONE
```

/sig storage

pohly committed:
e2e/storage: parameterize container images

The CSI integration test for hostpath was hard-coded to use the latest
stable release of the sidecar and hostpath container images. This
makes sense for regression testing of changes made in Kubernetes
itself, but the same test is also useful for testing the "canary"
images on quay.io before tagging them as a new release or for testing
locally produced images. Both is now possible via command line
parameters.

Testing "canary" images on quay.io:

  go run hack/e2e.go -- --provider=local --test \
     --test_args="--ginkgo.focus=CSI.plugin.test.using.CSI.driver..hostPath -csiImageVersion=canary"

Testing local container images:

  # https://docs.docker.com/registry/deploying/
  docker run -d -p 5000:5000 --restart=always --name registry registry:2
  for i in driver-registrar drivers external-attacher external-provisioner; do
     make -C $i REGISTRY_NAME=localhost:5000 push
  done
  go run hack/e2e.go -- --provider=local --test \
     --test_args="--ginkgo.focus=CSI.plugin.test.using.CSI.driver..hostPath -csiImageVersion=canary -csiImageRegistry=localhost:5000"

pohly commented:
@lpabon: this is the e2e test change that I mentioned last week.

pohly commented:
``make -C $i REGISTRY_NAME=localhost:5000 push`` depends on the currently pending PRs linked to in kubernetes-csi/docs#23.

wgliang commented:
/ok-to-test

dims commented on test/e2e/storage/csi_objects.go:
```diff
@@ -37,12 +38,30 @@ import (
+var csiImageRegistry string
+
+func init() {
+	flag.StringVar(&csiImageVersion, "csiImageVersion", "", "overrides the default tag used for hostpathplugin/csi-attacher/csi-provisioner/driver-registrar images")
```
how would one specify different image versions for different images (like what's currently listed in 42-45 lines above)?

pohly commented:
Davanum Srinivas <notifications@github.com> writes:
> how would one specify different image versions for different images
> (like what's currently listed in 42-45 lines above)?

One can't. I had thought about it, but one either would have to use
multiple different parameters or a much more complex parameter syntax,
so I decided against supporting that because I can't think of a
situation where it would be useful.

For the "test quay.io canary" use case the version is the same for all
containers. When building locally one can also use the same version
name.

dims commented:
Ack. thanks @pohly

msau42 commented on test/e2e/storage/csi_objects.go:
```diff
@@ -37,12 +38,30 @@ import (
+var csiImageVersion string
+var csiImageRegistry string
+
+func init() {
```
Should these be defined in framework/test_context.go instead?

pohly replied:
I didn't know that there was some centralized argument handling. As this seems to be preferred, I'll update the PR to put the flags there.

pohly replied:
Done, see the separate commit cc2a5954acf570b7f919a9a7e274661cc830d3a3. I can squash if you want.

pohly committed:
e2e/storage: central argument handling

Putting the command line argument handling into the central test
context seems like the better solution, in particular considering that
argument handling might get changed in the future to use Viper.

pohly commented:
/test pull-kubernetes-integration
/test pull-kubernetes-e2e-gce-device-plugin-gpu
/test pull-kubernetes-e2e-kops-aws

msau42 commented on test/e2e/storage/csi_objects.go:
```diff
@@ -37,12 +37,23 @@ import (
-	csiExternalProvisionerImage string = "quay.io/k8scsi/csi-provisioner:v0.2.1"
-	csiDriverRegistrarImage     string = "quay.io/k8scsi/driver-registrar:v0.2.0"
-)
+var csiImageVersions = map[string]string{
```
Call this defaultImageVersions?

msau42 reviewed (commented):
So if I understand correctly, if I'm making a change in one of the CSI components, in order to test that, I need to rebuild/retag all the CSI components since the image versions have to all be the same?

davidz627 commented:
It looks like the version flag would not be expressive enough to specify the default image versions we have set now. So I don't see how "so I decided against supporting that because I can't think of a situation where it would be useful." (https://github.com/kubernetes/kubernetes/pull/63780#issuecomment-388773433) holds true.

What is the use case for changing the `TestContext.CSIImageVersion` flag? The test would be forced to run all sidecars in the same version like @msau42 mentioned. If its just useful for "canary" then maybe just a boolean flag for "canary" would work instead?

pohly commented:
Michelle Au <notifications@github.com> writes:
> So if I understand correctly, if I'm making a change in one of the CSI
> components, in order to test that, I need to rebuild/retag all the CSI
> components since the image versions have to all be the same?

Correct. Now that my PRs about "consistent Docker image creation" have
been merged into each of the four CSI repos, one can rebuild all of them
with:

   for i in driver-registrar/ external-attacher/ external-provisioner/ drivers/; do
       make -C $i REGISTRY_NAME=localhost:5000 push
   done

Note that new container images always get pushed with tag
"canary". Setting a release tag happens after testing the current set of
canary images.

pohly commented:
David Zhu <notifications@github.com> writes:
> It looks like the version flag would not be expressive enough to
> specify the default image versions we have set now. So I don't see how
> "so I decided against supporting that because I can't think of a
> situation where it would be useful."
> (https://github.com/kubernetes/kubernetes/pull/63780#issuecomment-388773433)
> holds true.

It's true because *I* really can't think of a use case for overriding
individual version numbers ;-} But perhaps I'm just missing
something. Can you explain a situation where it would be useful?

> What is the use case for changing the `TestContext.CSIImageVersion`
> flag?

Testing the "canary" images on quay.io. See my other comment about how
the release process currently works (push "canary" images, test them,
tag them without rebuilding - at least that's my understanding from what
I have been told).

> The test would be forced to run all sidecars in the same version like
> @msau42 mentioned. If its just useful for "canary" then maybe just a
> boolean flag for "canary" would work instead?

I find that too limited. When pushing to my own registry, I might have a
set of images with tag "pre-invasive-change-foo" and another set with
"foo-done" and then I might want to switch testing back and forth
between the two.

msau42 commented:
I primarily see two kinds of developers:
* CSI driver developers would mostly be interested in changing the image version for their driver
* Kubernetes CSI sidecar developers would mostly be interested in changing the image version for 1 or more sidecar components

Would it be easier for development purposes to just put the manifests under test/e2e/testing-manifests/storage-csi and modify images and versions there as needed?

pohly commented:
Michelle Au <notifications@github.com> writes:
> I primarily see two kinds of developers:
> * CSI driver developers would mostly be interested in changing the
> image version for their driver

Those developers can't use this test, can they? It's hard-coded to use
the example hostpath driver.

> * Kubernetes CSI sidecar developers would mostly be interested in
> changing the image version for 1 or more sidecar components

IMHO (see my comments in
https://github.com/kubernetes-csi/drivers/issues/81) it would be better
for those developers to have all sidecard container code in one repo.
Then "make push" would automatically push all four containers into a
local registry, using the same tag, in which case there would be no need
to have per-container version override.

But you are right, if we continue to have separate repos and developers
only want to rebuild+test on of them, then we need it.

> Would it be easier for development purposes to just put the manifests
> under test/e2e/testing-manifests/storage-csi and modify images and
> versions there as needed?

This looks a bit cumbersome to use to me. I also don't like that one has
to modify source code of Kubernetes. If one isn't careful, this might
end up in a commit.

msau42 commented:
As an end goal, I would like to see this CSI suite develop into something where it can run against any CSI driver.  But I think that would require CSI driver deployment specs to be an input, and not just the image name/version.

pohly commented:
Michelle Au <notifications@github.com> writes:
> As an end goal, I would like to see this CSI suite develop into
> something where it can run against any CSI driver.  But I think that
> would require CSI driver deployment specs to be an input, and not just
> the image name/version.

Agreed.

So where does that leave us with the current patch set? Do we want to
keep hostpath-based testing as a special case with easy-to-use command
line options to configure the testing, or switch over hostpath to
testing with a deployment spec?

msau42 commented:
I discussed a bit with @saad-ali.  For the short term, making the image version and registry names parameters is fine for now.  But long term, let's definitely look into making the CSI driver manifests a parameter instead so we can test any CSI driver.  When that happens, then we may no longer need the parameters for versions and registries.

msau42 commented:
But I think this still seems a bit difficult to use from a developer pov.  If I'm changing kubernetes-csi sidecars, then I need to also rebuild the hostpath driver image.  If I'm only changing the hostpath driver, then I need to rebuild kubernetes-csi components too.

pohly commented:
Michelle Au <notifications@github.com> writes:
> But I think this still seems a bit difficult to use from a developer
> pov.  If I'm changing kubernetes-csi sidecars, then I need to also
> rebuild the hostpath driver image.  If I'm only changing the hostpath
> driver, then I need to rebuild kubernetes-csi components too.

Last week in the CSI standup meeting it was decided to go ahead with
merging all of these components into one repo. So yes, developers need
to rebuild them, but doing so will be both simple (one command) and fast
(it is not that much code after all).

For details, see
https://docs.google.com/document/d/1xC0Cm11WTebT9m_9yZbfzud24GJRxlp55gdR5QaMHmM/edit

msau42 commented:
I mean the hostpath csi driver is still built separately than kubernetes-csi component.

What do you think if we make this PR parameterize kubernetes-csi components, and leave parameterizing CSI drivers, like hostpath, to a kubernetes csi plugin e2e framework?

pohly commented:
Michelle Au <notifications@github.com> writes:
> I mean the hostpath csi driver is still built separately than
> kubernetes-csi component.

No, it is meant to be in the same repo as the sidecar apps, because it
gets developed by the same team and the components get tested together.

> What do you think if we make this PR parameterize kubernetes-csi
> components, and leave parameterizing CSI drivers, like hostpath, to a
> kubernetes csi plugin e2e framework?

That doesn't make sense when considering the direction above.

msau42 commented:
Ok sounds good for now then. Once we develop a better testing framework for csi plugins, then I expect these hostpath plugin tests to be reworked and may not work with this option anymore.

msau42 commented:
/lgtm
/approve no-issue

msau42 commented:
/assign @saad-ali

saad-ali commented:
/approve

jberkus commented:
Escalating priority so that this merges.

/priority critical-urgent

jberkus commented:
/remove-priority important-soon
